discovery/linode/linode.go

```diff
@@ -0,0 +1,195 @@
+	linodeLabelPrivateIPv4 = linodeLabel + "private_ipv4"
+	linodeLabelPublicIPv4  = linodeLabel + "public_ipv4"
+	linodeLabelPublicIPv6  = linodeLabel + "public_ipv6"
+	linodeLabelRegion      = linodeLabel + "region"
+	linodeLabelType        = linodeLabel + "type"
+	linodeLabelStatus      = linodeLabel + "status"
+	linodeLabelTags        = linodeLabel + "tags"
+	separator              = ","
+)
+
+// DefaultSDConfig is the default Linode SD configuration.
+var DefaultSDConfig = SDConfig{
+	Port:             80,
+	RefreshInterval:  model.Duration(60 * time.Second),
+	HTTPClientConfig: config.DefaultHTTPClientConfig,
+}
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// SDConfig is the configuration for Linode based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+}
+
+// Name returns the name of the Config.
+func (*SDConfig) Name() string { return "linode" }
+
+// NewDiscoverer returns a Discoverer for the Config.
+func (c *SDConfig) NewDiscoverer(opts discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(c, opts.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery periodically performs Linode requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client *linodego.Client
+	port   int
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port: conf.Port,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "linode_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client := linodego.NewClient(
+		&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		},
+	)
+	client.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version))
+	d.client = &client
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"linode",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "Linode",
+	}
+
+	// gather all linode instances
+	instances, err := d.client.ListInstances(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// gather detailed IP address info for all IPs on all linode instances
```

Comment:
Comments should start with a capital letter and end with a full stop.

Reply:
Can we use the info from the instance endpoint and then check ourselves if the address is private? https://github.com/golang/go/commit/1c9e587b90172e7654db897d8c938ffc665e1673

Reply:
For the current usage here, that would indeed work (and how I originally wrote it). However, we have an upcoming VLAN product that is currently in beta, and so I figured using a Linode-provided IP address classification may help make things clearer. 

I haven't looked too closely at the proposed API changes for VLANs, but that can also be added in the future once it's stable and available for production usage, in my opinion. 

I'm happy to switch things up to using the single instances call and classifying the IP addresses ourselves 👍

Reply:
If there is more to come, this seems fine. It is only one call for all the IP addresses.

Reply:
Ok, I'll leave it as-is for now then. I'll be happy to help work on future contributions as more is added to the Linode API/go library, as well 👍